reentr.c

```diff
@@ -421,27 +408,31 @@ Perl_reentrant_retry(const char *f, ...)
     case KEY_gethostbyname:
     case KEY_endhostent:
 	{
+            char * host_addr;
+            Size_t asize;
+            char * host_name;
+            int anint;
 
 #    ifdef PERL_REENTRANT_MAXSIZE
 	    if (PL_reentrant_buffer->_hostent_size <=
 		PERL_REENTRANT_MAXSIZE / 2)
 #    endif
-	    {
+            RenewDouble(PL_reentrant_buffer->_hostent_buffer,
+                    &PL_reentrant_buffer->_hostent_size, char);
             switch (key) {
-            case KEY_gethostbyaddr:
-		    p0    = va_arg(ap, void *);
-		    asize = va_arg(ap, size_t);
+	        case KEY_gethostbyaddr:
+		    host_addr = va_arg(ap, char *);
+		    asize = va_arg(ap, Size_t);
 		    anint  = va_arg(ap, int);
-		    retptr = gethostbyaddr((Netdb_host_t) p0, (Netdb_hlen_t) asize, anint); break;
+		    retptr = gethostbyaddr(host_addr, asize, anint); break;
```

Comment:
The second parameter under (old) POSIX for `gethostbyaddr()` is a `socklen_t`, which may be smaller than `size_t`.

This may cause warnings in the call to gethostbyaddr() without the cast.